crates/burn-dataset/src/dataset/dataframe.rs

```diff
@@ -0,0 +1,390 @@
+use std::marker::PhantomData;
+
+use polars::frame::row::Row;
+use polars::prelude::*;
+use serde::{
+    de::{self, DeserializeOwned, Deserializer, SeqAccess, Visitor},
+    Deserialize, forward_to_deserialize_any,
+};
+use serde::de::DeserializeSeed;
+
+use crate::Dataset;
+
+type Result<T> = core::result::Result<T, DataframeDatasetError>;
+
+/// Error type for DataframeDataset
+#[derive(thiserror::Error, Debug)]
+pub enum DataframeDatasetError {
+    /// Error occurred during deserialization
+    #[error("{0}")]
+    Other(String),
+}
+
+impl de::Error for DataframeDatasetError {
+    fn custom<T: std::fmt::Display>(msg: T) -> Self {
+        DataframeDatasetError::Other(msg.to_string())
+    }
+}
+
+/// Enum to hold either a DataFrame or a LazyFrame
+pub enum Frame {
+    /// Represents a Polars Dataframe
+    Data(DataFrame),
+    /// Represents a Polars LazyFrame
+    Lazy(LazyFrame),
+}
```

Comment:
Let's remove Lazy for now to get the approval quicker (I suggested Lazy as well but it appears fetching records more complex and currently you're just returning None).